### PR TITLE
Fix column typecasting with duplicate column names for PostgreSQL

### LINKED ISSUE #48946: Regression in Postgres aggregate function data type inference/casting in Rails >= 7.0.5
### Steps to reproduce
```sql
-- Create Postgres database, required to reproduce
CREATE DATABASE type_inference_test;
```

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # gem "rails", github: "rails/rails", tag: "v7.0.4.3" # OK
  # gem "rails", github: "rails/rails", tag: "v7.0.5"   # Fail
  # gem "rails", github: "rails/rails", tag: "v7.0.7"   # Fail
  gem "rails", github: "rails/rails", branch: "main"    # Fail
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# The issue appears appears with Postgres, not with SQLite. I have not tested other adapters.
# ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
# N.b., in sqlite3 .pluck("sum(decimal_column)").first returns a Float instead of a BigDecimal. Though that arguably is a bug, AFAICT it is _not_ a regression.
ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "type_inference_test", encoding: "unicode")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :line_items, force: true do |t|
    t.integer :unit_count
    t.decimal :subtotal
  end
end

class LineItem < ActiveRecord::Base
end

class TypeInferenceBug < Minitest::Test
  def setup
    LineItem.create! unit_count: 5, subtotal: 47.98
    LineItem.create! unit_count: 1, subtotal: 13.13
  end

  # This works correctly
  def test_single_column
    assert_kind_of Integer, LineItem.sum(:unit_count)
    assert_kind_of Integer, LineItem.pluck("sum(unit_count)").first
    assert_kind_of BigDecimal, LineItem.sum(:subtotal)
    assert_kind_of BigDecimal, LineItem.pluck("sum(subtotal)").first
  end

  # All these fail starting in 7.0.5. ActiveRecord seems to apply the type of the last item to all numeric(?) columns.
  def test_two_column_decimal_last
    # Fails in 7.0.5-7.0.7, ok in 7.0.4.3
    assert_kind_of Integer, LineItem.pluck("sum(unit_count)", "sum(subtotal)").first.first
  end

  def test_two_column_integer_last
    # Fails in 7.0.5-7.0.7, ok in 7.0.4.3
    assert_kind_of BigDecimal, LineItem.pluck("sum(subtotal)", "sum(unit_count)").first.first
  end

  def test_same_problem_with_another_aggregate_function
    # Fails in 7.0.5-7.0.7, ok in 7.0.4.3
    assert_kind_of BigDecimal, LineItem.pluck("max(subtotal)", "max(unit_count)").first.first
  end
end

```

### Expected behavior
Type of values returned from `pluck("aggregate_function(column)",...)` depends on database's data type and does not change if additional values are also plucked. Up through Rails 7.0.4.3, this worked as expected.

### Actual behavior
When plucking sums (or other aggregate functions) of more than one numeric column, `pluck` coerces all numeric(?) values to the type of the last column, at least when using the postgresql adapter. In our use, this results in integers inconveniently cast to BigDecimals and (more perniciously) decimal values representing dollars and cents being truncated to integers—and the data type of existing columns of a `pluck` call changing based on order or from adding an additional column.

### System configuration
**Rails version**: 7.0.7

**Ruby version**: 3.1.1p18

### PR DESCRIPTION
**Pull request by Flixt**
Fixes incorrect typecasting that occurred when there were duplicate column names returned from PostgreSQL which is always the case when calculations/aggregates are use multiple times when no aliases are used. The `column_type` method was moved from private to public, and the typecasting logic was updated.

Fixes #48946.

### Additional information

There are several other places where the `ActiveRecord::Result#column_type` is used directly, but I'm not sure if accessing the column_type via index would be correct/possible there. To me it seems these places are even incorrect since https://github.com/rails/rails/pull/45783 because the the `types`/`column_types` potentially not only contains columns by name but also by index.
- https://github.com/rails/rails/blob/f6b987d521b9ad3f9938b40f039fbdda78058a06/activerecord/lib/active_record/querying.rb#L66-L66
- https://github.com/rails/rails/blob/f6b987d521b9ad3f9938b40f039fbdda78058a06/activerecord/lib/active_record/associations/join_dependency.rb#L127-L127
- https://github.com/rails/rails/blob/f6b987d521b9ad3f9938b40f039fbdda78058a06/activerecord/lib/active_record/relation/calculations.rb#L508-L508

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/test/cases/calculations_test.rb [hunk 1]
```diff
@@ -1225,6 +1225,20 @@ def test_pluck_functions_without_alias
...
+    NumericData.create!(temperature: 1.0, decimal_number_big_precision: 5.23)
+    NumericData.create!(temperature: 2.0, decimal_number_big_precision: 4.56)
+    result = NumericData.pluck("sum(temperature)", "sum(decimal_number_big_precision)")
+
+    if current_adapter?(:PostgreSQLAdapter)
+      assert_equal Float, result.first.first.class
+      assert_equal BigDecimal, result.first.second.class
+    else
+      assert_equal Float, result.first.first.class
+      assert_equal Integer, result.first.second.class
```
Comment (Flixt) [inline activerecord/test/cases/calculations_test.rb:1238]: I'm a bit puzzled and would have expected `sum(decimal_number_big_precision)` to be mapped to `Float`. This is unrelated to the the changes of this PR also mapped to `Integer` without the changes provided here.

File: activerecord/lib/active_record/result.rb [hunk 2]
```diff
@@ -130,6 +130,14 @@ def cast_values(type_overrides = {}) # :nodoc:
       end
     end
 
+    def column_type(name, index, type_overrides)
```
Comment (Flixt) [inline activerecord/lib/active_record/result.rb:133]: I someone has an idea to retrieve the correct type without spreading the fallback logic all over the place and making this method public, hints would be very welcome.

File: activerecord/test/cases/calculations_test.rb [hunk 3]
```diff
@@ -1202,29 +1202,32 @@ def test_pluck_functions_with_alias
...
-        ["5", "The Fifth Topic of the day"]
-      ]
-    else
-      [
-        [1, "The First Topic"], [2, "The Second Topic of the day"],
-        [3, "The Third Topic of the day"], [4, "The Fourth Topic of the day"],
-        [5, "The Fifth Topic of the day"]
-      ]
-    end
+    expected = [
```
Comment (Flixt) [inline activerecord/test/cases/calculations_test.rb:1205]: This reverts a change introduced in https://github.com/rails/rails/pull/42933 which I think was incorrect and although the test passed. The plucks should return the correct types and not use type of the last column returned.

File: activerecord/lib/active_record/result.rb (hunk 2 above)
Comment (ghiculescu) [reply activerecord/lib/active_record/result.rb:133]: I'm not sure it answers your question directly, but you should remove this from the public API.

```suggestion
    def column_type(name, index, type_overrides) # :nodoc:
```

**Comment by kamipo** [comment]
Closing in favor of #52015.
